### DISCUSSION TITLE
Deprecating Legacy Provider Properties, `networkChanged` event, and `send()` `net_version` support

### DISCUSSION THREAD

**Discussion by jiexi** [MetaMask/metamask-improvement-proposals#23] [Breaking Changes]
~~Currently the following properties are accessible on the MetaMask provider object:~~
~~* `window.ethereum.chainId`~~
~~* `window.ethereum.networkVersion`~~
~~* `window.ethereum.selectedAddress`~~

~~They allow for synchronous access to the last known value of these chain/network attributes. They are not part of the [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) spec and are not officially listed in our [provider API documentation ](https://docs.metamask.io/wallet/reference/provider-api/). Despite not being intended for public usage, these properties have been publicly available for some time and we can assume that there are some Dapps out there that now rely on these values being available on the MetaMask provider object. Given this, we should make a best effort to inform developers that usage of these properties is not supported, that these properties will be removed in the future, and what equivalents can be used instead.~~

~~Related to the above fields, we also plan on removing the following which have already been properly deprecated:~~
~~* `networkChanged` event~~
~~* `net_version` support in `window.ethereum.send()`~~

The above announcement can be ignored. As of September 6th 2024, we plan on continuing to support:
* `window.ethereum.chainId`
* `window.ethereum.networkVersion`
* `window.ethereum.selectedAddress`
* `networkChanged` event
* `net_version` support in `window.ethereum.send()`

In a few releases, you should stop seeing console warnings when accessing `window.ethereum.chainId`, `window.ethereum.networkVersion`, and `window.ethereum.selectedAddress`.

We still recommend the migrations specified in [the comment below](https://github.com/MetaMask/metamask-improvement-proposals/discussions/23#discussioncomment-7912806).

**Comment by Mattijah** [comment]
Before getting rid of these properties it should be make sure that `eth_requestAccounts` and others actually work when called. https://github.com/MetaMask/metamask-extension/issues/7221

**Reply by vandan** [reply]
Please confirm whether the guidance from @jiexi on [#7221](https://github.com/MetaMask/metamask-extension/issues/7221#issuecomment-1839783964) resolves the problem that had been reported there.

**Comment by vandan** [comment]
Adding some recommended alternatives for each property we intend to remove:

`window.ethereum.chainId` - use `window.ethereum.request({method: 'eth_chainId'})`
`window.ethereum.networkVersion` - if you actually need network version then use `window.ethereum.request({method: 'net_version'})` , otherwise use `eth_chainId`
`window.ethereum.selectedAddress` - use `window.ethereum.request({method: 'eth_accounts'})`
`networkChanged` event - no exact equivalent, use the `chainChanged` event instead
`net_version` support in `window.ethereum.send()` - use `window.ethereum.request({method: 'net_version'})`

**Reply by u1x30** [reply]
To check if a MetaMask account is locked without triggering the metamask unlock popup, we currently use window.ethereum.selectedAddress. However, using window.ethereum.request({ method: 'eth_accounts' }) initiates the MetaMask popup to unlock account, and we want to avoid that. Anybody know if there an alternative method to determine if the account is locked and has an address without triggering the MetaMask popup?

**Reply by jiexi** [reply]
When called with MetaMask locked, `window.ethereum.request({ method: 'eth_accounts' })` does not initiate a prompt from MetaMask to unlock the account for me. Are you able to provide steps to reproduce this behavior?

`eth_accounts` will return `[]` if no accounts are connected or if MetaMask is locked which should be equivalent to the `window.ethereum.selectedAddress` behavior.

**Comment by SaveSnowflake** [comment]
After one page updated and added new wallets top their page , now my hot wallet refuses to launch on it or when it does - it shows the issue,  so i check the dev tools to find the exact issue
![Image 16-2-24 at 16 17](https://github.com/MetaMask/metamask-improvement-proposals/assets/88808051/7bfbc22a-c453-48c7-b05d-83130d112b48)

**Comment by Number000001** [comment]
With the recommended “window.ethereum.request({ method: ‘net_version’ })”, “await ethereum.request({ method: ‘eth_chainId’ });”, “await web3.eth.getChainId();”, and “await web3.eth.net.getId();”, if I change the network after connecting MetaMask to a DApp, only the network ID at the time of connection is returned, not the changed network ID.

As far as I know, if you change the network after connecting MetaMask to DApps, only “window.ethereum.networkVersion” can get the changed network ID, so I really hope it won’t be deleted.

With “window.ethereum.on(‘chainChanged’, async (chainId) =>”, I can obtain the changed network ID even if I change the network after connecting MetaMask to a DApp, but this is inconvenient, so “window.ethereum.networkVersion”, which allows me to actively obtain the changed network ID, is more convenient.

If I change the network after connecting MetaMask to a DApp, is there any other way to obtain the new network ID?

Thanks.

**Comment by elix1er** [comment]
bump

**Comment by Ryan-yang125** [comment]
Still MetaMask: 'ethereum.selectedAddress' is deprecated and may be removed in the future. Please use the 'eth_accounts' RPC method instead.
Anything update?

**Comment by stanleyseow** [comment]
How do I get rid of this metamask errors from the console ??

**Comment by yawayl** [comment]
to be honest i dont know what im doing but i do wanna know how im ruining everything for everyone else, please some one take heed as im getting upset with this crypto world. ive  learned only street value online so not tech savy. i know this web world is as real as we can make it and so far only i think i know %100 isReal to my benifit.  one thing for sure is this that i feel like i am making all these bugs that every a/i is loving, especailly the sweeper bug, it was force upon me and now im using it to clean house the web house.  im probly nobody but now no body has a hold of every body god in particaler, escuse my illiterate, thats key role online lol. reach out some one please be persistent i go by the handle ghost or godhost. ty

**Reply by nichechristie** [reply]
Who is this

**Comment by jiexi** [comment]
Please see the updated announcement. Sorry for any inconvenience the original announcement may have caused.

**Comment by Madeindreams** [comment]
Did anyone tried having Metamask, Phantom, Coinbase wallets installed in the same browser?

On eth_requestAccounts. you first get to select phantom or metamask.
After that you get to select coinbase or metamask.
then after all that, when you selected metamask.

The request fail with

MetaMask connection failed: RangeError: Maximum call stack size exceeded


Hopefuly this will all be resolved with EIP-6963

**Reply by vandan** [reply]
You can learn more about MetaMask's wallet interoperability support [here](https://docs.metamask.io/wallet/concepts/wallet-interoperability/#eip-6963-interfaces). EIP-6963 support was implemented and has been available since last year, but each dapp must also implement EIP-6963 to allow multiple extension wallets to be used on their site without errors like this.

**Reply by Madeindreams** [reply]
Yes but I guess it also depend on the browser! I was made aware of this because Brave nightly came out with new attribute like 
isBrave, IsMetamask. And window.etherum was not working as expected. Will get back to it's implementation soon. Thank you.

**Comment by Silvrbckw** [comment]
Can you share with those of us who don't know: where do we retrieve the Eth
Accounts RPC? Thanks.
www.rcs-distribution.online

**Comment by Silvrbckw** [comment]
What is the ETA of EIP-6963? Thanks.
www.rcs-distribution.online

**Reply by vandan** [reply]
As mentioned above, EIP-6963 support has been available in MetaMask since late 2023, but dapps must implement EIP-6963 on their end as well. Many popular dapps already support EIP-6963.

**Comment by yawayl** [comment]
Why is hacking a positive initiation enhaenced becoming negative.  can reversal loop around now Era, please.  I am just a partical that fell on a dust atom which nuetralizes the H20 thats very salty as the city that conjuered each crystal yet waiting to be proccess and end up on a table for your liking. Then again who cares about the Ocean.  I just hope the fuss of an bug like covid -18.755 example of how i got to null this digi-man world is still att hello.  one houndred per sent  verifing honest is prime.  so Are we there yet beez. what is Sat. Nok. or me all day System Naming. lol lets begin with words beings caused it weighs oddly upon each shoulder that is a11even? only my bard nolls this Mass aged.  Thank you for looking.
